flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-postgres/src/main/java/org/apache/flink/cdc/connectors/postgres/source/reader/PostgresPipelineRecordEmitter.java

```diff
@@ -0,0 +1,129 @@
+            PostgresDialect postgresDialect) {
+        super(
+                debeziumDeserializationSchema,
+                sourceReaderMetrics,
+                sourceConfig.isIncludeSchemaChanges(),
+                offsetFactory);
+        this.sourceConfig = sourceConfig;
+        this.postgresDialect = postgresDialect;
+        this.alreadySendCreateTableTables = new HashSet<>();
+        this.createTableEventCache = new ArrayList<>();
+
+        if (!sourceConfig.getStartupOptions().startupMode.equals(StartupMode.INITIAL)) {
+            try (PostgresConnection jdbc = postgresDialect.openJdbcConnection()) {
+                List<TableId> capturedTableIds =
+                        TableDiscoveryUtils.listTables(
+                                sourceConfig.getDatabaseList().get(0),
+                                jdbc,
+                                sourceConfig.getTableFilters());
+                for (TableId tableId : capturedTableIds) {
+                    Schema schema = PostgresSchemaUtils.getTableSchema(tableId, sourceConfig, jdbc);
+                    createTableEventCache.add(
+                            new CreateTableEvent(
+                                    org.apache.flink.cdc.common.event.TableId.tableId(
+                                            tableId.schema(), tableId.table()),
+                                    schema));
+                }
+            } catch (SQLException e) {
+                throw new RuntimeException("Cannot start emitter to fetch table schema.", e);
+            }
+        }
+    }
+
+    @Override
+    protected void processElement(
+            SourceRecord element, SourceOutput<T> output, SourceSplitState splitState)
+            throws Exception {
+        if (isLowWatermarkEvent(element) && splitState.isSnapshotSplitState()) {
+            TableId tableId = splitState.asSnapshotSplitState().toSourceSplit().getTableId();
+            if (!alreadySendCreateTableTables.contains(tableId)) {
+                try (PostgresConnection jdbc = postgresDialect.openJdbcConnection()) {
+                    sendCreateTableEvent(jdbc, tableId, (SourceOutput<Event>) output);
+                    alreadySendCreateTableTables.add(tableId);
+                }
+            }
+        } else if (splitState.isStreamSplitState()
+                && !alreadySendCreateTableForBinlogSplit
+                && !sourceConfig.getStartupOptions().startupMode.equals(StartupMode.INITIAL)) {
+            for (CreateTableEvent createTableEvent : createTableEventCache) {
+                output.collect((T) createTableEvent);
+            }
+            alreadySendCreateTableForBinlogSplit = true;
+        }
+        super.processElement(element, output, splitState);
+    }
+
+    private void sendCreateTableEvent(
+            PostgresConnection jdbc, TableId tableId, SourceOutput<Event> output) {
+        Schema schema = PostgresSchemaUtils.getTableSchema(tableId, sourceConfig, jdbc);
```

Comment:
What if log's schema is different with current table schema?

Reply:
Hello, I looked at the code and found that debezium queries the table schema by calling postgresql-42.5.1.jar

![image](https://github.com/apache/flink-cdc/assets/125547374/720a61a4-fff0-4580-a44e-8d4c95138cc4)

![image](https://github.com/apache/flink-cdc/assets/125547374/158f0e37-de8b-4379-82ab-6ac6cc6fd9bc)

![image](https://github.com/apache/flink-cdc/assets/125547374/827ef7f6-ef15-439a-af87-e753a75d5900)